### PR TITLE
[pipeline] Statistic pipeline's schedule overhead

### PR DESCRIPTION
**Pull request by Pslydhh**
- Count the thread level scheduling overhead for each fragment of each query

### CODE REVIEW

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 1]
```diff
@@ -45,20 +57,33 @@ void GlobalDriverDispatcher::finalize_driver(DriverRawPtr driver, RuntimeState*
...
         }
 
         size_t queue_index;
+
+        timespec start;
+        clock_gettime(CLOCK_MONOTONIC, &start);
         auto maybe_driver = this->_driver_queue->take(&queue_index);
+        timespec end;
+        clock_gettime(CLOCK_MONOTONIC, &end);
+
```
Comment (liuyehcf) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:73]: Could it be replaced with ScopedTimer or MonotonicStopWatch?

File: be/src/exec/pipeline/pipeline_driver_dispatcher.cpp [hunk 2]
```diff
@@ -45,20 +57,33 @@ void GlobalDriverDispatcher::finalize_driver(DriverRawPtr driver, RuntimeState*
...
+void GlobalDriverDispatcher::run(size_t thread_id) {
     while (true) {
         if (_num_threads_setter.should_shrink()) {
             break;
         }
 
         size_t queue_index;
+
+        timespec start;
+        clock_gettime(CLOCK_MONOTONIC, &start);
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.cpp:69]: std::chrono::high_resolution_clock？

File: be/src/exec/pipeline/fragment_context.h [hunk 3]
```diff
@@ -36,6 +36,25 @@ class FragmentContext {
...
+            if (milliseconds >= config::pipeline_thread_schedule_threshold) {
+                has_thread_shedule_overhead = true;
+                os << "THREAD " << i;
+                os << ": " << _thread_shedule_frequency[i] << "-" << milliseconds << "ms(" << (milliseconds / 1000)
+                   << "s) " << ((milliseconds * 1000) / _thread_shedule_frequency[i]) << "us/frequency"
+                   << "; ";
+            }
+        }
+
+        if (has_thread_shedule_overhead) {
```
Comment (kangkaisen) [inline be/src/exec/pipeline/fragment_context.h:53]: Needn't log it, profile is enough

File: be/src/exec/pipeline/fragment_context.h [hunk 4]
```diff
@@ -151,6 +177,10 @@ class FragmentContext {
     std::atomic<Status*> _final_status;
     std::atomic<bool> _cancel_flag;
     Status _s_status;
+
+public:
+    std::vector<int64_t> _thread_shedule_time;
```
Comment (kangkaisen) [inline be/src/exec/pipeline/fragment_context.h:182]: Why need a vector? I think one value is enough

File: be/src/exec/pipeline/pipeline_driver_dispatcher.h [hunk 5]
```diff
@@ -57,6 +61,34 @@ class GlobalDriverDispatcher final : public FactoryMethod<DriverDispatcher, Glob
...
+
+    class ThreadIdAllocator {
+    public:
+        ThreadIdAllocator() : upper_bound(0), freed(), lock() {}
+        size_t allocate() {
+            std::lock_guard<std::mutex> guard(lock);
+            if (freed.empty()) {
+                return upper_bound++;
+            } else {
+                size_t id = freed.top();
```
Comment (kangkaisen) [inline be/src/exec/pipeline/pipeline_driver_dispatcher.h:73]: Why need to reuse the thread_id?

**Pull request closed without merging**